compiler/rustc_const_eval/src/check_consts/ops.rs

```diff
@@ -355,20 +357,82 @@ fn build_error_for_const_call<'tcx>(
                 non_or_conditionally,
             })
         }
-        _ => ccx.dcx().create_err(errors::NonConstFnCall {
-            span,
-            def_descr: ccx.tcx.def_descr(callee),
-            def_path_str: ccx.tcx.def_path_str_with_args(callee, args),
-            kind: ccx.const_kind(),
-            non_or_conditionally,
-        }),
+        _ => {
+            let def_descr = ccx.tcx.def_descr(callee);
+            let mut err = ccx.dcx().create_err(errors::NonConstFnCall {
+                span,
+                def_descr,
+                def_path_str: ccx.tcx.def_path_str_with_args(callee, args),
+                kind: ccx.const_kind(),
+                non_or_conditionally,
+            });
+            let context_span = ccx.tcx.def_span(ccx.def_id());
+            err.span_label(context_span, format!(
+                "calls in {}s are limited to constant functions, tuple structs and tuple variants",
+                ccx.const_kind(),
+            ));
+            note = false;
+            let def_kind = ccx.tcx.def_kind(callee);
+            if let DefKind::AssocTy | DefKind::AssocConst | DefKind::AssocFn = def_kind {
```

Comment:
Callees are always `AssocFn`s, why did you add `DefKind::AssocTy | DefKind::AssocConst` too?